d3log/src/transact.rs

```diff
@@ -0,0 +1,130 @@
+use std::time::{SystemTime,UNIX_EPOCH};
+use std::collections::HashMap;
+use crate::{tcp_network::TcpNetwork,
+            batch::Batch,
+            Node};
+use mm_ddlog::api::HDDlog;
+use std::sync::Arc;
+//use std::sync::Mutex as SyncMutex;
+use async_std::sync::Mutex;
+use tokio::task;
+use differential_datalog::{DeltaMap, D3log, program::Update, DDlog, DDlogDynamic};
+
+pub type Timestamp = u64;
+
+
+pub struct TransactionManager {
+    // svcc needs the membership, so we are going to assign nids
+    progress : Vec<Timestamp>,
+    
+    // would like tcp network to be a trait object
+    n : TcpNetwork,
+    
+    // need access to some hddlog to call d3log_localize_val
+    h : HDDlog,
+}
+// ok, Instant is monotonic, but SystemTime is not..we need something
+// which is going to track across restarts and as monotonic, so we
+// will burtally coerce SytemTime into monotonicity
+
+fn current_time() ->Timestamp {
+    let now = SystemTime::now();
+    let delta = now
+        .duration_since(UNIX_EPOCH)
+        // fix this with a little sequence number
+        .expect("Monotonicity violation");
+    let ms = delta.subsec_millis();
+    return ms as u64;
+}
+
+impl TransactionManager {
+    fn start(){
+    }
+    
+    // pass network
+    pub fn new() -> TransactionManager {
+        match HDDlog::run(1, false) {
+            Ok((hddlog, _init_output)) => {
+                TransactionManager{
+                    n:TcpNetwork::new(),
+                    h:hddlog, // arc?
+                    progress:Vec::new(),
+                }
+            }
+            Err(err) => {
+                // this is supposed to be management protocol
+                panic!("Failed to run differential datalog: {}", err);
+            }
+        }
+    }
+}
+
+    
+// no asynch trait
+// maybe this belongs in network?
+
+pub async fn forward(tm : Arc::<Mutex::<TransactionManager>>,  input: Batch)  -> Result<(), std::io::Error> {
```

Comment:
This should be a method of `TransactionManager` even if the first argument is not called `self`.